_source/_posts/2024-07-05-jhipster-terraform-azure.md

```diff
@@ -0,0 +1,1857 @@
+
+{% img blog/jhipster-terraform-azure/azure-architecture.jpg alt:"Azure architecture diagram" width:"900" %}{: .center-image }
+
+**Network Topology**: The architecture uses a hub-spoke network topology, where the hub and spoke are deployed in separate virtual networks connected through peering, minimizing exposure of resources to the public internet. The hub virtual network contains an Azure Firewall in a dedicated subnet, securing outbound network traffic. The spoke network contains the AKS cluster, the Application Gateway, and a private link to the Azure Container Registry.
+
+**Azure Kubernetes Service**: The managed Kubernetes cluster.
+
+**Azure Virtual Network**: an isolated and secure environment for VMs and applications.
+
+**Azure Application Gateway**: Load balances traffic to the web application, operating at Layer 7, using the Azure Application Gateway Ingress Controller (AGIC) as the Kubernetes ingress controller. It has the Web Application Firewall (WAF) enabled, securing incoming traffic from common web attacks and it can perform SSL termination.
+
+> **IMPORTANT NOTE**: To limit the scope of this tutorial, TLS is not configured at the Application Gateway, but keep in mind securing traffic to your application is required for production.
+
+**Azure Firewall**: the network security service protecting all network resources, only allowing approved outbound traffic, by configuring firewall rules.
+
+**Azure Container Registry**: stores private container images that can be run in the AKS cluster.
+
+**Application Gateway Ingress Controller (AGIC)**: a Kubernetes resource for leveraging the Azure Application Gateway as an external load balancer for exposing an application to the internet.
+
+On to the deployment, start by creating a folder for the Terraform configuration, inspired by the [reference implementation](https://github.com/mspnp/aks-fabrikam-dronedelivery) from Azure:
+
+```shell
+mkdir terraform
+cd terraform
+mkdir modules
+touch main.tf
+touch outputs.tf
+touch variables.tf
+touch providers.tf
+```
+
+Edit the file `providers.tf` and add the following content:
+
+```terraform
+# terraform/providers.tf
+terraform {
+  required_version = ">=1.8"
+
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "~>3.107"
+    }
+    random = {
+      source  = "hashicorp/random"
+      version = "~>3.0"
+    }
+    azapi = {
+      source  = "azure/azapi"
+      version = "~>1.5"
+    }
+  }
+}
+
+provider "azurerm" {
+  features {}
+}
+```
+
+### Configure a Hub Network and Azure Firewall
+
+Create a module for the hub network configuration:
+
+```shell
+cd modules
+mkdir hub_network
+cd hub_network
+touch main.tf
+touch outputs.tf
+touch variables.tf
+```
+
+Edit `main.tf` and add the following content:
+
+```terraform
+# terraform/modules/hub_network/main.tf
+locals {
```

Comment:
@indiepopart do you thibnk we can make this script shorter by removing fields that are optional and has a default value? This is too long for a blog IMO

Reply:
Did a split and removed some optionals.